packages/test/test-service-load/src/loadTestDataStore.ts

```diff
@@ -473,31 +484,56 @@ class LoadTestDataStore extends DataObject implements ILoadTest {
+        const opSizeinBytes = (typeof config.testConfig.opSizeinBytes === 'undefined') ? 
+        0 : config.testConfig.opSizeinBytes;
+        assert(opSizeinBytes >= 0, "opSizeinBytes must be greater than or equal to zero.");
+        const generateStringOfSize = (sizeInBytes: number): string => new Array(sizeInBytes + 1).join("0");
+        let opsSent = 0; 
         try {
-            while (dataModel.counter.value < clientSendCount && !this.disposed) {
-                // this enables a quick ramp down. due to restart, some clients can lag
-                // leading to a slow ramp down. so if there are less than half the clients
-                // and it's partner is done, return true to complete the runner.
-                if (this.runtime.getAudience().getMembers().size < config.testConfig.numClients / 2
-                    && ((await dataModel.getPartnerCounter())?.value ?? 0) >= clientSendCount) {
-                    return true;
+            if (opSizeinBytes === 0) {
+                while (dataModel.counter.value < clientSendCount && !this.disposed) {
+                    // this enables a quick ramp down. due to restart, some clients can lag
+                    // leading to a slow ramp down. so if there are less than half the clients
+                    // and it's partner is done, return true to complete the runner.
+                    if (this.runtime.getAudience().getMembers().size < config.testConfig.numClients / 2
+                        && ((await dataModel.getPartnerCounter())?.value ?? 0) >= clientSendCount) {
+                        return true;
+                    }
+                    if (dataModel.assigned()) {
+                        dataModel.counter.increment(1);
+                        opsSent = dataModel.counter.value;
+                        if (opsSent % opsPerCycle === 0) {
+                            await dataModel.blobFinish();
+                            dataModel.abandonTask();
+                            // give our partner a half cycle to get the task
+                            await delay(cycleMs / 2);
+                        } else {
+                            // Random jitter of +- 50% of opWaitMs
+                            await delay(opsGapMs + opsGapMs * random.real(0, .5, true)(config.randEng));
+                        }
+                    } else {
+                        await dataModel.volunteerForTask();
+                    }
                 }
-
-                if (dataModel.assigned()) {
-                    dataModel.counter.increment(1);
-                    if (dataModel.counter.value % opsPerCycle === 0) {
-                        await dataModel.blobFinish();
-                        dataModel.abandonTask();
-                        // give our partner a half cycle to get the task
-                        await delay(cycleMs / 2);
+            } else {
+                while (opsSent < clientSendCount && !this.disposed) {
```

Comment:
it's a little awkward that so much of this code needs to be duplicated. I'm not very familiar with the concerns of our service load tests, but maybe one path to making this better would be to abstract the operations currently done on counter (i.e. get its value and increment) into the test data store, and instead set "counter + string" objects into the sharedmap when `opSizeInBites > 0`. This is making a tighter commitment to keep the lack of concurrency in operations to the counter/map though, since map has last-write-win semantics.

There are some other directions you could take; I just would love to see less duplication in this area.

Reply:
note: my suggested approach would probably not work well with #12901. at the very least, concurrent ops could cause observed counter values to be lower than the actual number of semantic increments. So if that's important for load testing, I'd do something else. There are probably a lot of reasonable solutions, I'm just not particularly familiar with the setup of these tests and what they're optimized for. E.g. converting the old codepath to also be based on the number of sent ops (tracked locally) might be the right path instead. I just like to see complexity removed wherever possible.

Reply:
Hey @Abe27342, it is a great suggestion in terms of keeping the code cohesive. It can surely be taken into as an action plan to consider all the functionalities we support currently and come up with a more optimal solution of representing that.